SonarQube.Client/Requests/AggregatingRequestFactory.cs

```diff
@@ -0,0 +1,45 @@
+﻿/*
+ * SonarQube Client
+ * Copyright (C) 2016-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarQube.Client.Requests
+{
+    internal class AggregatingRequestFactory : IRequestFactory
+    {
+        private readonly IRequestFactory[] requestFactories;
+
+        public AggregatingRequestFactory(params IRequestFactory[] requestFactories)
+        {
+            this.requestFactories = requestFactories;
+        }
+
+        public TRequest Create<TRequest>(ServerInfo serverInfo) where TRequest : class, IRequest
+        {
+            foreach (var factory in requestFactories)
```

Comment:
I'm a bit confused why would we need this?
The server is always either SC or SQ.
This seems to attempt first call SQ, and only if it fails, call SC (or vice-versa). That doesn't seem right.
We should be calling whatever server is configured and only that one. If the API is not available in it, we should fail, not fallback to a different server API. It's also entirely possible that the API would be available in both versions but return different values.

I don't think we should be iterating over the factories.
We should have only 2, get one of them at the start of the connection, and then use only that one until the connection is reset.

Reply:
Yep, you're right. That's what I originally intended to do, but managed to confuse myself into thinking it wouldn't work without more refactoring. However, I've just been looking at `ConnectAsync` and it should be doable.

FYI in the aggregating approach the factory implementations would have checked the server type and returned null if they didn't handle the server type.